File: src/modelo/ProdutoServico.java
package modelo;

import pojo.*;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * SERVIÇO 1 — ProdutoServico
 * Classe de modelo que implementa operações de negócio sobre produtos.
 * Simula o que seria uma camada de serviço (Service Layer).
 */
public class ProdutoServico {

    // "banco de dados" em memória para testes
    private List<Produto> repositorio = new ArrayList<>();
    private int proximoId = 1;

    // ── CRUD básico ──────────────────────────────────────────────────────────

    public Produto cadastrar(Produto p) {
        p.setId(proximoId++);
        repositorio.add(p);
        System.out.println("[SERVIÇO] Produto cadastrado: " + p);
        return p;
    }

    public Produto buscarPorId(int id) {
        return repositorio.stream()
                          .filter(p -> p.getId() == id)
                          .findFirst()
                          .orElse(null);
    }

    public List<Produto> listarTodos() {
        return new ArrayList<>(repositorio);
    }

    public boolean remover(int id) {
        boolean ok = repositorio.removeIf(p -> p.getId() == id);
        if (ok) System.out.println("[SERVIÇO] Produto id=" + id + " removido.");
        return ok;
    }

    // ── Regras de negócio ────────────────────────────────────────────────────

    /**
     * Lista apenas vacinas perecíveis que estão vencidas.
     */
    public List<VacinaPerecivel> listarVencidasPerecíveis() {
        return repositorio.stream()
                .filter(p -> p instanceof VacinaPerecivel)
                .map(p -> (VacinaPerecivel) p)
                .filter(VacinaPerecivel::isVencido)
                .collect(Collectors.toList());
    }

    /**
     * Lista todos os produtos por espécie alvo.
     * Só funciona para ProdutoVeterinario e subclasses.
     */
    public List<ProdutoVeterinario> buscarPorEspecie(String especie) {
        return repositorio.stream()
                .filter(p -> p instanceof ProdutoVeterinario)
                .map(p -> (ProdutoVeterinario) p)
                .filter(pv -> pv.getEspecieAlvo().equalsIgnoreCase(especie))
                .collect(Collectors.toList());
    }

    /**
     * Calcula o valor total do estoque (soma dos preços).
     */
    public double calcularValorTotal() {
        return repositorio.stream()
                          .mapToDouble(Produto::getPreco)
                          .sum();
    }

    /**
     * Serializa um produto para String no formato chave=valor.
     * Usado antes de enviar via Stream/TCP.
     */
    public String serializar(Produto p) {
        StringBuilder sb = new StringBuilder();
        sb.append("id=").append(p.getId()).append(";");
        sb.append("nome=").append(p.getNome()).append(";");
        sb.append("preco=").append(p.getPreco()).append(";");
        sb.append("fabricante=").append(p.getFabricante()).append(";");
        sb.append("tipo=").append(p.getClass().getSimpleName());
        return sb.toString();
    }

    public int getTotalProdutos() { return repositorio.size(); }
}

File: vet/cliente/modelo/ProdutoServico.cpp
#include "ProdutoServico.hpp"

ProdutoServico::ProdutoServico() {}

Produto* ProdutoServico::cadastrar(Produto* p) {
    p->setId(proximoId++);
    repositorio.push_back(p);
    cout << "[SERVICO] Produto cadastrado: " << p->getNome() << endl;
    return p;
}

Produto* ProdutoServico::buscarPorId(int id) {
    for (auto* produto : repositorio) {
        if (produto->getId() == id) {
            return produto;
        }
    }
    return nullptr;
}

vector<Produto*> ProdutoServico::listarTodos() const {
    return this->repositorio;
}

bool ProdutoServico::remover(int id) {
    for (auto it = this->repositorio.begin(); it != this->repositorio.end(); ++it) {
        if ((*it)->getId() == id) {
            delete *it; // libera a memoria do produto removido
            this->repositorio.erase(it);
            cout << "[SERVIÇO] Produto id=" << id << " removido." << endl;
            return true;
        }
    }
    return false;
}

vector<VacinaPerecivel*> ProdutoServico::listarVacinasPereciveis() const {
    vector<VacinaPerecivel*> vacinas;
    for (auto* produto : repositorio) {
        if (auto* vacina = dynamic_cast<VacinaPerecivel*>(produto)) {
            if (vacina->ehVencido())
            {
                vacinas.push_back(vacina);
            }
            
        }
    }
    return vacinas;
}

string ProdutoServico::serializar(Produto* p) const {
    string result = "";

    result += "id=" + to_string(p->getId()) + ";";
    result += "nome=" + p->getNome() + ";";
    result += "preco=" + to_string(p->getPreco()) + ";";
    result += "fabricante=" + p->getFabricante() + ";";

    result += "tipo=" + string(typeid(*p).name()) + ";";
    return result;
}

int ProdutoServico::getTotalProdutos() const {
    return this->repositorio.size();
}

vector<ProdutoVeterinario*> ProdutoServico::buscarPorEspecie(const string& especie) const {
    vector<ProdutoVeterinario*> produtosEncontrados;
    for (auto* produto : repositorio) {
        if (auto* prodVet = dynamic_cast<ProdutoVeterinario*>(produto)) {
            if (prodVet->getEspecieAlvo() == especie) {
                produtosEncontrados.push_back(prodVet);
            }
        }
    }
    return produtosEncontrados;
}

double ProdutoServico::calcularValorTotal() const {
    double total = 0.0;
    for (auto* produto : repositorio) {
        total += produto->getPreco();
    }
    return total;
}

ProdutoServico::~ProdutoServico() {
    for (auto* produto : repositorio) {
        delete produto;
    }
}